torch/_inductor/ir.py

```diff
@@ -5567,3 +5593,71 @@ def codegen_collective(self, wrapper, output_name, input_names):
             f"group={output_name}_pg, "
             "async_op=True)"
         )
+
+
+class AllToAllSingle(CollectiveKernelAlloc):
+    """
+    NOTE: `all_to_all_single` is implemented differently from other collectives
+    because its output shape is data-dependent on input `output_split_sizes`.
+    Instead of letting Inductor manage allocation of output buffer that has
+    data-dependent dim size which is difficult to do, we resort to using
+    `ExternKernelAlloc` and letting the underlying `fun_col_impl._all_to_all_single`
```

Comment:
i'm confused how this is different from other out-of-place collectives? for those, inductor also does not alloc and relies on c10d to perform the allocation.  the 'inplace' ones are special in that they hijack an input buffer created by inductor and let that be used in place.

Reply:
we seem to have created a chimera in 'CollectiveKernel' since like ExternKernelAlloc, it implements 'should_allocate' to return False, but unlike ExternKernelAlloc, it will fail certain isinstance checks scattered through inductor.  This probably isn't good.  Did you hit one of those cases and hence needed to subclass ExternKernelAlloc?  Im wondering if we should have made 'OutOfPlaceCollectiveKernel' subclass from ExternKernelAlloc, but didn't since that opens the door to multiple-inheritance problems.  cc @kumpera

Reply:
thanks - moved back to subclass from `OutOfPlaceCollectiveKernel`. Found mainly 3 differences compared to other existing out-of-place collectives:
1. `create_output_buffers` is not called, because we rely on the underlying `fun_col_impl._all_to_all_single` kernel to do the output buffer allocation.
2. `fun_col_impl._register_tensor_work` is not emitted, because the underlying `fun_col_impl._all_to_all_single` kernel already calls it.
3. `c10d._find_or_create_pg_by_ranks_and_tag` is not emitted, because the underlying `fun_col_impl._all_to_all_single` kernel already calls it.

(documented in code as well)